cpp/include/cuspatial/experimental/detail/pairwise_point_in_polygon.cuh

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/error.hpp>
+#include <cuspatial/experimental/detail/is_point_in_polygon.cuh>
+#include <cuspatial/traits.hpp>
+#include <cuspatial/vec_2d.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <thrust/memory.h>
+
+#include <iterator>
+#include <type_traits>
+
+namespace cuspatial {
+namespace detail {
+
+template <class Cart2dItA,
+          class Cart2dItB,
+          class OffsetIteratorA,
+          class OffsetIteratorB,
+          class OutputIt,
+          class Cart2dItADiffType = typename std::iterator_traits<Cart2dItA>::difference_type,
+          class Cart2dItBDiffType = typename std::iterator_traits<Cart2dItB>::difference_type,
+          class OffsetItADiffType = typename std::iterator_traits<OffsetIteratorA>::difference_type,
+          class OffsetItBDiffType = typename std::iterator_traits<OffsetIteratorB>::difference_type>
+__global__ void pairwise_point_in_polygon_kernel(Cart2dItA test_points_first,
+                                                 Cart2dItADiffType const num_test_points,
+                                                 OffsetIteratorA poly_offsets_first,
+                                                 OffsetItADiffType const num_polys,
+                                                 OffsetIteratorB ring_offsets_first,
+                                                 OffsetItBDiffType const num_rings,
+                                                 Cart2dItB poly_points_first,
+                                                 Cart2dItBDiffType const num_poly_points,
+                                                 OutputIt result)
+{
+  using Cart2d     = iterator_value_type<Cart2dItA>;
+  using OffsetType = iterator_value_type<OffsetIteratorA>;
+  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x;
+       idx <
+       std::distance(test_points_first, thrust::prev(test_points_first + num_test_points + 1));
+       idx += gridDim.x * blockDim.x) {
```

Comment:
```suggestion
  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x; idx < num_test_points; idx += gridDim.x * blockDim.x) {
```

Reply:
OMG I totally missed that in my comment. :)

Reply:
Done